dashboard/app/models/email_preference.rb

```diff
@@ -17,4 +17,23 @@
 #
 
 class EmailPreference < ApplicationRecord
+  validates_presence_of :email, :ip_address, :source
+  validates_inclusion_of :opt_in, {in: [true, false]}
+  validates_uniqueness_of :email
+  validates_email_format_of :email
+
+  def email=(value)
+    super(value&.strip&.downcase)
+  end
+
+  def self.upsert!(email, opt_in, ip_address, source, form_kind)
```

Comment:
with this many params, it could be nice to use keyword arguments.

Reply:
Done!